File: app/src/main/java/com/kinwatt/powermeter/sensor/bluetooth/GattClientBase.java
package com.kinwatt.powermeter.sensor.bluetooth;

import android.bluetooth.BluetoothDevice;
import android.bluetooth.BluetoothGatt;
import android.bluetooth.BluetoothGattCallback;
import android.bluetooth.BluetoothGattCharacteristic;
import android.bluetooth.BluetoothGattDescriptor;
import android.bluetooth.BluetoothGattService;
import android.bluetooth.BluetoothProfile;
import android.content.Context;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

public abstract class GattClientBase {
    private static final String TAG = GattClientBase.class.getSimpleName();

    private static final UUID CLIENT_CHARACTERISTIC_CONFIG = UUID.fromString("00002902-0000-1000-8000-00805F9B34FB");
    private static final UUID SERVER_CHARACTERISTIC_CONFIG = UUID.fromString("00002903-0000-1000-8000-00805F9B34FB");

    private Context context;

    private UUID serviceUuid;
    private List<UUID> notifications = new ArrayList<>();

    private BluetoothDevice device;
    private BluetoothGatt gatt;
    private BluetoothGattService service;

    private List<NotificationListener> listeners = new ArrayList<>();

    public GattClientBase(Context context, BluetoothDevice device, UUID serviceUuid) {
        this.context = context;
        this.device = device;
        this.serviceUuid = serviceUuid;
    }

    public void enableNotifications(UUID uuid) {
        if (service == null) {
            notifications.add(uuid);
        }
        else {
            enableNotificationsInternal(uuid);
        }
    }

    public void connect() {
        gatt = device.connectGatt(context, false, callback);

        if (gatt == null) {
            Log.i(TAG, "Unable to connect GATT server");
        }
        else {
            Log.i(TAG, "Trying to connect GATT server");
        }
    }

    public void close() {
        gatt.close();
        gatt = null;
        service = null;
    }

    public void addListener(NotificationListener listener) {
        listeners.add(listener);
    }
    public void removeListener(NotificationListener listener) {
        listeners.remove(listener);
    }

    protected void onNotificationReceived(Characteristic characteristic) {
        for (NotificationListener listener : listeners) {
            listener.onNotificationReceived(characteristic);
        }
    }

    protected abstract Characteristic decode(BluetoothGattCharacteristic characteristic);

    private void enableNotificationsInternal(UUID uuid) {
        enableNotificationsInternal(service.getCharacteristic(uuid));
    }

    private void enableNotificationsInternal(BluetoothGattCharacteristic characteristic) {
        if (gatt.setCharacteristicNotification(characteristic, true)) {
            BluetoothGattDescriptor descriptor = characteristic.getDescriptor(CLIENT_CHARACTERISTIC_CONFIG);
            descriptor.setValue(BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE);
            gatt.writeDescriptor(descriptor);
        }
    }

    private BluetoothGattCallback callback = new BluetoothGattCallback() {

        @Override
        public void onConnectionStateChange(BluetoothGatt gatt, int status, int newState) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Log.i(TAG, "BluetoothDevice CONNECTED: " + gatt.getDevice().getAddress() + " " + gatt.getDevice().getName());
                gatt.discoverServices();
            }
            else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Log.i(TAG, "BluetoothDevice DISCONNECTED");
                if (status == 133) {
                    Log.i(TAG, "Many connections");
                }
            }
        }

        @Override
        public void onServicesDiscovered(BluetoothGatt gatt, int status) {
            Log.i(TAG, "Services discovered");

            service = gatt.getService(serviceUuid);

            for (UUID uuid : notifications) {
                enableNotificationsInternal(uuid);
            }
        }

        @Override
        public void onCharacteristicChanged(BluetoothGatt gatt, BluetoothGattCharacteristic characteristic) {
            if (notifications.contains(characteristic.getUuid())) {
                onNotificationReceived(decode(characteristic));
            }
        }
    };
}

File: app/src/main/java/com/kinwatt/powermeter/sensor/bluetooth/GattClientBase.kt
package com.kinwatt.powermeter.sensor.bluetooth

import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothGatt
import android.bluetooth.BluetoothGattCallback
import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattDescriptor
import android.bluetooth.BluetoothGattService
import android.bluetooth.BluetoothProfile
import android.content.Context
import android.util.Log

import java.util.ArrayList
import java.util.UUID

abstract class GattClientBase(private val context: Context, private val device: BluetoothDevice, private val serviceUuid: UUID) {
    private val notifications = ArrayList<UUID>()
    private var gatt: BluetoothGatt? = null
    private var service: BluetoothGattService? = null

    private val listeners = ArrayList<NotificationListener>()

    private val callback = object : BluetoothGattCallback() {

        override fun onConnectionStateChange(gatt: BluetoothGatt, status: Int, newState: Int) {
            if (newState == BluetoothProfile.STATE_CONNECTED) {
                Logger.info("BluetoothDevice CONNECTED: " + gatt.device.address + " " + gatt.device.name)
                gatt.discoverServices()
            } else if (newState == BluetoothProfile.STATE_DISCONNECTED) {
                Logger.info("BluetoothDevice DISCONNECTED")
                if (status == 133) {
                    Logger.info("Many connections")
                }
            }
        }

        override fun onServicesDiscovered(gatt: BluetoothGatt, status: Int) {
            Logger.info("Services discovered")

            service = gatt.getService(serviceUuid)

            for (uuid in notifications) {
                enableNotificationsInternal(uuid)
            }
        }

        override fun onCharacteristicChanged(gatt: BluetoothGatt, characteristic: BluetoothGattCharacteristic) {
            if (notifications.contains(characteristic.uuid)) {
                onNotificationReceived(decode(characteristic))
            }
        }
    }

    fun enableNotifications(uuid: UUID) {
        if (service == null) {
            notifications.add(uuid)
        } else {
            enableNotificationsInternal(uuid)
        }
    }

    fun connect() {
        gatt = device.connectGatt(context, false, callback)

        if (gatt == null) {
            Logger.info("Unable to connect GATT server")
        } else {
            Logger.info("Trying to connect GATT server")
        }
    }

    fun close() {
        gatt!!.close()
        gatt = null
        service = null
    }

    fun addListener(listener: NotificationListener) {
        listeners.add(listener)
    }

    fun removeListener(listener: NotificationListener) {
        listeners.remove(listener)
    }

    protected open fun onNotificationReceived(characteristic: Characteristic) {
        for (listener in listeners) {
            listener.onNotificationReceived(characteristic)
        }
    }

    protected abstract fun decode(characteristic: BluetoothGattCharacteristic): Characteristic

    private fun enableNotificationsInternal(uuid: UUID) {
        enableNotificationsInternal(service!!.getCharacteristic(uuid))
    }

    private fun enableNotificationsInternal(characteristic: BluetoothGattCharacteristic) {
        if (gatt!!.setCharacteristicNotification(characteristic, true)) {
            val descriptor = characteristic.getDescriptor(CLIENT_CHARACTERISTIC_CONFIG)
            descriptor.value = BluetoothGattDescriptor.ENABLE_NOTIFICATION_VALUE
            gatt!!.writeDescriptor(descriptor)
        }
    }

    companion object {
        protected val Logger = java.util.logging.Logger.getLogger(GattClientBase::class.java.simpleName)!!

        private val CLIENT_CHARACTERISTIC_CONFIG = UUID.fromString("00002902-0000-1000-8000-00805F9B34FB")
        private val SERVER_CHARACTERISTIC_CONFIG = UUID.fromString("00002903-0000-1000-8000-00805F9B34FB")
    }
}

